File: frontend/src/utils/geo.js
/**
 * Shared geographic utility functions.
 */

const R_NM = 3440.065 // Earth radius in nautical miles

/**
 * Great-circle distance in nautical miles. Returns null if any argument is null/undefined.
 */
export function haversineNm(lat1, lon1, lat2, lon2) {
  if ([lat1, lon1, lat2, lon2].some(v => v == null)) return null
  const dLat = (lat2 - lat1) * Math.PI / 180
  const dLon = (lon2 - lon1) * Math.PI / 180
  const a = Math.sin(dLat / 2) ** 2
    + Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) * Math.sin(dLon / 2) ** 2
  return R_NM * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))
}

/**
 * True bearing in degrees (0=N, 90=E) from (lat1,lon1) to (lat2,lon2).
 */
export function bearingDeg(lat1, lon1, lat2, lon2) {
  const dLon = (lon2 - lon1) * Math.PI / 180
  const lat1r = lat1 * Math.PI / 180
  const lat2r = lat2 * Math.PI / 180
  const x = Math.sin(dLon) * Math.cos(lat2r)
  const y = Math.cos(lat1r) * Math.sin(lat2r) - Math.sin(lat1r) * Math.cos(lat2r) * Math.cos(dLon)
  return (Math.atan2(x, y) * 180 / Math.PI + 360) % 360
}

File: backend/radar/geo.py
from __future__ import annotations

import math as _math


def _bearing_deg_simple(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Forward azimuth from point 1 to point 2 in [0, 360)."""
    phi1 = _math.radians(lat1)
    phi2 = _math.radians(lat2)
    dlam = _math.radians(lon2 - lon1)
    x = _math.cos(phi1) * _math.sin(phi2) - _math.sin(phi1) * _math.cos(phi2) * _math.cos(dlam)
    y = _math.sin(dlam) * _math.cos(phi2)
    return (_math.degrees(_math.atan2(y, x)) + 360) % 360


def _haversine_nm_simple(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Great-circle distance between two lat/lon points in nautical miles."""
    r_m = 6_371_000.0
    phi1 = _math.radians(lat1)
    phi2 = _math.radians(lat2)
    dphi = _math.radians(lat2 - lat1)
    dlam = _math.radians(lon2 - lon1)
    a = _math.sin(dphi / 2.0) ** 2 + _math.cos(phi1) * _math.cos(phi2) * _math.sin(dlam / 2.0) ** 2
    c = 2.0 * _math.asin(_math.sqrt(max(0.0, min(1.0, a))))
    return (r_m * c) / 1852.0
